docs/source/developing-volttron/developing-agents/agent-development.rst

```diff
@@ -629,6 +629,115 @@ The VOLTTRON team has come up with a number of methods to help users develop mor
+
+ Type-hints do not validate the type of data input to a function call, they are merely suggestions about what the
+ function is expecting.
+
+* To specify the type expected as input:
+
+.. code-block:: python
+
+   # When calling this RPC method, the user should supply a string as input
+   @RPC.export
+   def type_hint_rpc(input_string: str):
+
+* To specify the type of function output:
+
+.. code-block:: python
+
+   # This demonstrates a function that expects a string as input and that will return an integer value
+   @RPC.export
+   def type_hint_rpc(input_string: str) -> int:
+
+* Specifying multiple types:
+
+.. code-block:: python
+
+    # Here our function expects either a string or dictionary
+    @RPC.export
+    def type_hint_rpc(input_json: Union[str, dict]) -> str:
+
+* To specify an optional argument with None as the default value:
+
+.. code-block:: python
+
+    # 'Optional' is used to specify either a string should be passed or the default value 'None' will be used
+    @RPC.export
+    def type_hint_rpc(optional_input: Optional[str] = None) -> str:
+
+* These techniques can be combined:
+
+.. code-block:: python
+
+    # 'Optional' can be used in combination with 'Union' for optional arguments which expect one of multiple types and
+    # default to None
+    @RPC.export
+    def type_hint_rpc(optional_input: Optional[Union[str, dict]] = None) -> str:
+
+
+API Validation
+""""""""""""""
+
+Each function within an agent should validate its input parameters, especially with structured data.
+
+* Make use of isinstance to do type checking:
+
+    .. code-block:: python
+
+       @RPC.export
+       def type_checking_rpc(input_str: str) -> dict:
+           if not isinstance(input_str, str):
+               # Include a descriptive error message to help the user determine why input validation failed
+               # You can make use of 'f-strings' to help the user with debugging
+               raise ValueError(
+                   f'The expected input type for function "type_checking_rpc" is str, received {type(input_str)}')
+
+* Add logic to validate the range of values supplied as input with a valid type:
+
+    .. code-block:: python
+
+       @RPC.export
+       def value_checking_rpc(input_json: Union[str, dict]) -> dict:
+           if not isinstance(input_json, str) or not isinstance(input_json, dict):
+               # You can make use of 'f-strings' to help the user determine why input validation failed
+               raise ValueError(
+                   f'The expected input type for function "type_checking_rpc" is str or dict, received {type(input_str)}')
+           else:
+               # since we expected the input to be valid JSON, be sure that it can be correctly parsed
+               if isinstance(input_json, str):
+                   input_json = json.loads(input_json)
+               # for this example, we expect our JSON to include two fields: test1 and test2
+               test_1 = input_json.get("test1")
```

Comment:
This might lead to KeyError if key is not present. Explain about putting it in try-except block or assigning a default value. test_1 = input_json.get("test1", "")